src/pepr/loki/index.ts

```diff
@@ -0,0 +1,107 @@
+import * as yaml from "js-yaml";
+import { Capability, a } from "pepr";
+import { Component, setupLogger } from "../logger";
+
+const log = setupLogger(Component.LOKI);
+
+export const loki = new Capability({
+  name: "loki",
+  description: "UDS Core Capability for the Loki stack.",
+});
+
+const { When } = loki;
+
+// Define the type for the YAML configuration
+interface SchemaConfig {
+  from: string;
+  store: string;
+  object_store: string;
+  schema: string;
+  index: {
+    prefix: string;
+    period: string;
+  };
+}
+
+interface LimitsConfig {
+  allow_structured_metadata: boolean;
+}
+
+interface LokiConfig {
+  auth_enabled: boolean;
+  chunk_store_config: object;
+  common: object;
+  frontend: object;
+  frontend_worker: object;
+  index_gateway: object;
+  limits_config: LimitsConfig;
+  memberlist: object;
+  pattern_ingester: object;
+  query_range: object;
+  ruler: object;
+  runtime_config: object;
+  schema_config: {
+    configs: SchemaConfig[];
+  };
+  server: object;
+  storage_config: object;
+  tracing: object;
+}
+
+When(a.Secret)
+  .IsCreatedOrUpdated()
+  .WithLabel("app.kubernetes.io/instance", "loki")
+  .WithLabel("app.kubernetes.io/name", "loki")
+  .Mutate(async secret => {
+    log.info(
+      secret,
+      `Processing Secret ${secret.Raw.metadata?.namespace}/${secret.Raw.metadata?.name} for loki schemaconfig date updates.`,
+    );
+
+    if (secret.Raw.data && secret.Raw.data["config.yaml"]) {
+      // Parse the config.yaml content into a LokiConfig object
+      let lokiConfig: LokiConfig;
+      try {
+        lokiConfig = yaml.load(secret.Raw.data["config.yaml"]) as LokiConfig;
+      } catch (e) {
+        log.error(secret, `Failed to parse Loki config.yaml: ${e.message}`);
+        return;
+      }
+
+      // Check if schema_config and configs exist
+      if (lokiConfig.schema_config && Array.isArray(lokiConfig.schema_config.configs)) {
+        // Get the v13 schema configuration
+        const v13Config = lokiConfig.schema_config.configs.find(config => config.schema === "v13");
+
+        if (v13Config) {
+          const currentDate = new Date();
+
+          if (new Date(v13Config.from) < currentDate) {
+            // Calculate the new date 2 days in advance
+            currentDate.setDate(currentDate.getDate() + 2);
+
+            // Update the v13 schema configuration date
+            v13Config.from = currentDate.toISOString().split("T")[0]; // Format as YYYY-MM-DD
+
+            // Ensure limits_config exists and update allow_structured_metadata to false
+            if (!lokiConfig.limits_config) {
+              lokiConfig.limits_config = {} as LimitsConfig;
+            }
+            lokiConfig.limits_config.allow_structured_metadata = false;
+
+            // Update the secret with the new config.yaml content
+            secret.Raw.data["config.yaml"] = yaml.dump(lokiConfig);
+            log.info(secret.Raw.data["config.yaml"], `Secret config.yaml updated successfully.`);
```

Comment:
I think there might be some missing logic here - but want to summarize my understanding of the cases and how we should be handling them:
- install/upgrade: the date (as set in this PR) would be a previous date (2024-06-04) and we would want to generate a future date, 2 days in advance as discussed to be safe with the way Loki handles this
- future upgrade: the date coming into the webhook would be our previous date (2024-06-04), we would want to set this to the date we already generated (probably could query the cluster for this?)

The key here is that we want to generate a date in the future once, but once we generate it once always use that date for future mutations (since the baseline applied from the chart will always be an older date).

Reply:
Yeah that is a good point that i totally misunderstood. I've updated so the date is only generated once and stored as an annotation in the secret for future reference.